server/backends/invocationdb/invocationdb.go

```diff
@@ -53,17 +53,27 @@ func (d *InvocationDB) createInvocation(tx *db.DB, ctx context.Context, ti *tabl
 	return tx.Create(ti).Error
 }
 
-func (d *InvocationDB) InsertOrUpdateInvocation(ctx context.Context, ti *tables.Invocation) error {
-	return d.h.Transaction(ctx, func(tx *db.DB) error {
+// InsertOrUpdateInvocation checks whether an invocation with the same primary
+// key as the given invocation exists. If no such invocation exists, it will
+// create a new row with the given invocation properties. Otherwise, it will
+// update the existing invocation. It returns whether a new row was created.
+func (d *InvocationDB) InsertOrUpdateInvocation(ctx context.Context, ti *tables.Invocation) (bool, error) {
+	created := false
+	err := d.h.Transaction(ctx, func(tx *db.DB) error {
 		var existing tables.Invocation
 		if err := tx.Where("invocation_id = ?", ti.InvocationID).First(&existing).Error; err != nil {
 			if !db.IsRecordNotFound(err) {
 				return err
 			}
-			return d.createInvocation(tx, ctx, ti)
+			if err := d.createInvocation(tx, ctx, ti); err != nil {
+				created = true
```

Comment:
The returned object from GORM's `Create` method contains [`RowsAffected`](https://gorm.io/docs/create.html#Create-Record), which would possibly be a better indicator for if we created a row than the value of the Error object is. Maybe change `createInvocation` to also return `RowsAffected`?

Reply:
If the invocation already exists when calling createInvocation, then gorm should return an AlreadyExists error, so I think checking the error is OK

Reply:
sounds good; just wanted to bring up the option.